libs/mpi/include/hpx/mpi/mpi_future.hpp

```diff
@@ -0,0 +1,515 @@
+            MPI_Comm_size(MPI_COMM_WORLD, &detail::mpi_info_.size_);
+        }
+        else if (init_mpi) {
+            int required, provided;
+            required = MPI_THREAD_MULTIPLE;
+            MPI_Init_thread(0, nullptr, required, &provided);
+            if (provided < MPI_THREAD_FUNNELED) {
+                mpi_debug.error(debug::str<>("hpx::mpi::init"), "init failed");
+                throw std::runtime_error(
+                    "Your MPI installation doesn't allow multiple threads");
+            }
+            MPI_Comm_rank(MPI_COMM_WORLD, &detail::mpi_info_.rank_);
+            MPI_Comm_size(MPI_COMM_WORLD, &detail::mpi_info_.size_);
+            detail::mpi_info_.mpi_initialized_ = true;
+        }
+        mpi_debug.debug(debug::str<>("hpx::mpi::init"), detail::mpi_info_);
+
+        if (init_errorhandler) {
+            set_error_handler();
+            detail::mpi_info_.error_handler_initialized_ = true;
+        }
+
+        // install polling loop on requested thread pool
+        if (pool_name.empty())
+        {
+            detail::register_polling(hpx::resource::get_thread_pool(0));
+        }
+        else
+        {
+            detail::register_polling(hpx::resource::get_thread_pool(pool_name));
+        }
+    }
+
+    // -----------------------------------------------------------------
+    namespace detail
+    {
+        void unregister_polling(hpx::threads::thread_pool_base& pool)
+        {
+            auto* sched = pool.get_scheduler();
+            sched->remove_scheduler_mode(threads::policies::enable_user_polling);
+        }
+    }    // namespace detail
+
+    void finalize(std::string const& pool_name = "")
+    {
+        if (detail::mpi_info_.mpi_initialized_) {
+            MPI_Finalize();
+        }
+        if (detail::mpi_info_.error_handler_initialized_) {
+            mpi_debug.debug("error handler deletion not implemented");
+        }
+        //
+        mpi_debug.debug(debug::str<>("Clearing mode")
+              , detail::mpi_info_, "disable_user_polling");
+        if (pool_name.empty())
+        {
+            detail::unregister_polling(hpx::resource::get_thread_pool(0));
+        }
+        else
+        {
+            detail::unregister_polling(
+                hpx::resource::get_thread_pool(pool_name));
+        }
+    }
+
+    // -----------------------------------------------------------------
+    // This RAII helper class assumes that MPI initialization/finalization is
+    // handled elsewhere
+    struct enable_user_polling
+    {
+        enable_user_polling(std::string const& pool_name = "")
+          : pool_name_(pool_name)
+        {
+            mpi::init(false, pool_name);
+        }
+
+        ~enable_user_polling()
+        {
+            mpi::finalize(pool_name_);
+        }
+
+        template <typename F>
+        void wait(F&& f)
+        {
+            return mpi::wait(std::forward<F>(f));
+        }
```

Comment:
I know, I added this member function `wait` here. I think we should not have that, however. Let's not advertise the existence of a global wait facility.